Index: JavaScriptCore/runtime/DateConstructor.cpp
===================================================================
--- JavaScriptCore/runtime/DateConstructor.cpp	(revision 41760)
+++ JavaScriptCore/runtime/DateConstructor.cpp	(working copy)
@@ -78,7 +78,10 @@
         else {
             JSValuePtr primitive = args.at(exec, 0).toPrimitive(exec);
             if (primitive.isString())
-                value = parseDate(primitive.getString());
+                // If we are preventing user profiling, then we need to ensure that if
+                // a timezone is not specified we don't use local time. This would allow
+                // later calls to getTimeZoneOffset() to reveal the user's actual timezone.
+                value = parseDate(primitive.getString(), exec->dynamicGlobalObject()->shouldPreventUserProfiling());
             else
                 value = primitive.toNumber(exec);
         }
@@ -102,7 +105,10 @@
           t.second = args.at(exec, 5).toInt32(exec);
           t.isDST = -1;
           double ms = (numArgs >= 7) ? args.at(exec, 6).toNumber(exec) : 0;
-          value = gregorianDateTimeToMS(t, ms, false);
+          // If we are preventing user profiling, then we need to ensure that if
+          // a timezone is not specified we don't use local time. This would allow
+          // later calls to getTimeZoneOffset() to reveal the user's actual timezone.
+          value = gregorianDateTimeToMS(t, ms, exec->dynamicGlobalObject()->shouldPreventUserProfiling());
         }
     }
 
Index: JavaScriptCore/runtime/DateMath.h
===================================================================
--- JavaScriptCore/runtime/DateMath.h	(revision 41760)
+++ JavaScriptCore/runtime/DateMath.h	(working copy)
@@ -60,7 +60,7 @@
 void getLocalTime(const time_t*, tm*);
 
 // Not really math related, but this is currently the only shared place to put these.  
-double parseDate(const UString&);
+double parseDate(const UString&, bool treatLocalAsUTC = false);
 double timeClip(double);
 UString formatDate(const GregorianDateTime&);
 UString formatDateUTCVariant(const GregorianDateTime&);
@@ -80,7 +80,7 @@
 const double msPerDay = 24.0 * 60.0 * 60.0 * 1000.0;
 
 // Intentionally overridding the default tm of the system
-// Tee members of tm differ on various operating systems.
+// The members of tm differ on various operating systems.
 struct GregorianDateTime : Noncopyable {
     GregorianDateTime()
         : second(0)
Index: JavaScriptCore/runtime/DateMath.cpp
===================================================================
--- JavaScriptCore/runtime/DateMath.cpp	(revision 41760)
+++ JavaScriptCore/runtime/DateMath.cpp	(working copy)
@@ -596,7 +596,7 @@
     return true;
 }
 
-double parseDate(const UString &date)
+double parseDate(const UString &date, bool treatLocalAsUTC)
 {
     // This parses a date in the form:
     //     Tuesday, 09-Nov-99 23:12:40 GMT
@@ -879,7 +879,7 @@
         t.hour = hour;
 
         // Use our gregorianDateTimeToMS() rather than mktime() as the latter can't handle the full year range.
-        return gregorianDateTimeToMS(t, 0, false);
+        return gregorianDateTimeToMS(t, 0, treatLocalAsUTC);
     }
 
     return (ymdhmsToSeconds(year, month + 1, day, hour, minute, second) - (offset * 60.0)) * msPerSecond;
Index: JavaScriptCore/runtime/DatePrototype.cpp
===================================================================
--- JavaScriptCore/runtime/DatePrototype.cpp	(revision 41760)
+++ JavaScriptCore/runtime/DatePrototype.cpp	(working copy)
@@ -359,7 +359,6 @@
     : DateInstance(structure)
 {
     setInternalValue(jsNaN(exec));
-    // The constructor will be added later, after DateConstructor has been built.
 }
 
 bool DatePrototype::getOwnPropertySlot(ExecState* exec, const Identifier& propertyName, PropertySlot& slot)
@@ -369,11 +368,16 @@
 
 // Functions
 
-JSValuePtr dateProtoFuncToString(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList&)
+JSValuePtr dateProtoFuncToString(ExecState* exec, JSObject* result, JSValuePtr thisValue, const ArgList& args)
 {
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // If we are preventing user profiling, then always return UTC time.
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling()) {
+        return dateProtoFuncToUTCString(exec, result, thisValue, args);
+    }
+
     const bool utc = false;
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
@@ -408,7 +412,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -425,7 +430,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -437,11 +443,16 @@
     return jsNontrivialString(exec, formatTime(t, utc));
 }
 
-JSValuePtr dateProtoFuncToLocaleString(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
+JSValuePtr dateProtoFuncToLocaleString(ExecState* exec, JSObject* result, JSValuePtr thisValue, const ArgList& args)
 {
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // If we are preventing user profiling, then always return UTC time.
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling()) {
+        return dateProtoFuncToUTCString(exec, result, thisValue, args);
+    }
+
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
@@ -450,11 +461,16 @@
     return formatLocaleDate(exec, thisDateObj, milli, LocaleDateAndTime, args);
 }
 
-JSValuePtr dateProtoFuncToLocaleDateString(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
+JSValuePtr dateProtoFuncToLocaleDateString(ExecState* exec, JSObject* result, JSValuePtr thisValue, const ArgList& args)
 {
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // If we are preventing user profiling, then always return UTC time.
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling()) {
+        return dateProtoFuncToDateString(exec, result, thisValue, args);
+    }
+
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
@@ -463,11 +479,16 @@
     return formatLocaleDate(exec, thisDateObj, milli, LocaleDate, args);
 }
 
-JSValuePtr dateProtoFuncToLocaleTimeString(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
+JSValuePtr dateProtoFuncToLocaleTimeString(ExecState* exec, JSObject* result, JSValuePtr thisValue, const ArgList& args)
 {
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // If we are preventing user profiling, then always return UTC time.
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling()) {
+        return dateProtoFuncToTimeString(exec, result, thisValue, args);
+    }
+
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
@@ -494,7 +515,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -545,7 +567,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -579,7 +602,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -613,7 +637,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -647,7 +672,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -681,9 +707,10 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
-    DateInstance* thisDateObj = asDateInstance(thisValue); 
+    DateInstance* thisDateObj = asDateInstance(thisValue);
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
         return jsNaN(exec);
@@ -715,7 +742,8 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
-    const bool utc = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool utc = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
@@ -744,11 +772,16 @@
     return jsNumber(exec, t.second);
 }
 
-JSValuePtr dateProtoFuncGetMilliSeconds(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList&)
+JSValuePtr dateProtoFuncGetMilliSeconds(ExecState* exec, JSObject* result, JSValuePtr thisValue, const ArgList& argList)
 {
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // If we are preventing user profiling, then always return UTC time.
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling()) {
+        return dateProtoFuncGetUTCMilliseconds(exec, result, thisValue, argList);
+    }
+
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
@@ -764,6 +797,7 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+
     DateInstance* thisDateObj = asDateInstance(thisValue); 
     double milli = thisDateObj->internalNumber();
     if (isnan(milli))
@@ -779,6 +813,10 @@
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    if (exec->dynamicGlobalObject()->shouldPreventUserProfiling())
+        return jsNumber(exec,0);
+
     const bool utc = false;
 
     DateInstance* thisDateObj = asDateInstance(thisValue); 
@@ -874,7 +912,8 @@
 
 JSValuePtr dateProtoFuncSetMilliSeconds(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromTimeArgs(exec, thisValue, args, 1, inputIsUTC);
 }
 
@@ -886,7 +925,8 @@
 
 JSValuePtr dateProtoFuncSetSeconds(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromTimeArgs(exec, thisValue, args, 2, inputIsUTC);
 }
 
@@ -898,7 +938,8 @@
 
 JSValuePtr dateProtoFuncSetMinutes(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromTimeArgs(exec, thisValue, args, 3, inputIsUTC);
 }
 
@@ -910,7 +951,8 @@
 
 JSValuePtr dateProtoFuncSetHours(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromTimeArgs(exec, thisValue, args, 4, inputIsUTC);
 }
 
@@ -922,7 +964,8 @@
 
 JSValuePtr dateProtoFuncSetDate(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromDateArgs(exec, thisValue, args, 1, inputIsUTC);
 }
 
@@ -934,7 +977,8 @@
 
 JSValuePtr dateProtoFuncSetMonth(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromDateArgs(exec, thisValue, args, 2, inputIsUTC);
 }
 
@@ -946,7 +990,8 @@
 
 JSValuePtr dateProtoFuncSetFullYear(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList& args)
 {
-    const bool inputIsUTC = false;
+    // If we are preventing user profiling, then always return UTC time.
+    const bool inputIsUTC = exec->dynamicGlobalObject()->shouldPreventUserProfiling();
     return setNewValueFromDateArgs(exec, thisValue, args, 3, inputIsUTC);
 }
 
Index: JavaScriptCore/runtime/JSGlobalObject.h
===================================================================
--- JavaScriptCore/runtime/JSGlobalObject.h	(revision 41760)
+++ JavaScriptCore/runtime/JSGlobalObject.h	(working copy)
@@ -231,6 +231,8 @@
 
         virtual bool shouldInterruptScript() const { return true; }
 
+        virtual bool shouldPreventUserProfiling() const { return true; }
+        
         virtual bool allowsAccessFrom(const JSGlobalObject*) const { return true; }
 
         virtual bool isDynamicScope() const;
Index: WebKit/mac/WebCoreSupport/WebInspectorClient.mm
===================================================================
--- WebKit/mac/WebCoreSupport/WebInspectorClient.mm	(revision 41760)
+++ WebKit/mac/WebCoreSupport/WebInspectorClient.mm	(working copy)
@@ -167,6 +167,7 @@
     WebPreferences *preferences = [[WebPreferences alloc] init];
     [preferences setAutosaves:NO];
     [preferences setPrivateBrowsingEnabled:YES];
+    [preferences setPreventUserProfiling:NO];
     [preferences setLoadsImagesAutomatically:YES];
     [preferences setAuthorAndUserStylesEnabled:YES];
     [preferences setJavaScriptEnabled:YES];
Index: WebKit/mac/WebView/WebPreferences.h
===================================================================
--- WebKit/mac/WebView/WebPreferences.h	(revision 41760)
+++ WebKit/mac/WebView/WebPreferences.h	(working copy)
@@ -366,6 +366,19 @@
 - (BOOL)privateBrowsingEnabled;
 
 /*!
+    @method setPreventUserProfiling:
+    @param flag 
+    @abstract Prevents websites from using the 'visited' property
+        in CSS stylesheets to gather information on your browsing history..
+ */
+- (void)setPreventUserProfiling:(BOOL)flag;
+
+/*!
+    @method privateBrowsingEnabled
+ */
+- (BOOL)preventUserProfiling;
+
+/*!
     @method setTabsToLinks:
     @param flag 
     @abstract If tabsToLinks is YES, the tab key will focus links and form controls. 
Index: WebKit/mac/WebView/WebPreferenceKeysPrivate.h
===================================================================
--- WebKit/mac/WebView/WebPreferenceKeysPrivate.h	(revision 41760)
+++ WebKit/mac/WebView/WebPreferenceKeysPrivate.h	(working copy)
@@ -60,6 +60,7 @@
 #define WebKitBackForwardCacheExpirationIntervalKey @"WebKitBackForwardCacheExpirationIntervalKey"
 #define WebKitTabToLinksPreferenceKey @"WebKitTabToLinksPreferenceKey"
 #define WebKitPrivateBrowsingEnabledPreferenceKey @"WebKitPrivateBrowsingEnabled"
+#define WebKitPreventUserProfilingPreferenceKey @"WebKitPreventUserProfiling"
 #define WebContinuousSpellCheckingEnabled @"WebContinuousSpellCheckingEnabled"
 #define WebGrammarCheckingEnabled @"WebGrammarCheckingEnabled"
 #define WebKitDOMPasteAllowedPreferenceKey @"WebKitDOMPasteAllowedPreferenceKey"
Index: WebKit/mac/WebView/WebView.mm
===================================================================
--- WebKit/mac/WebView/WebView.mm	(revision 41760)
+++ WebKit/mac/WebView/WebView.mm	(working copy)
@@ -1324,6 +1324,7 @@
     settings->setDatabasesEnabled([preferences databasesEnabled]);
     settings->setLocalStorageEnabled([preferences localStorageEnabled]);
     settings->setPrivateBrowsingEnabled([preferences privateBrowsingEnabled]);
+    settings->setPreventUserProfiling([preferences preventUserProfiling]);
     settings->setSansSerifFontFamily([preferences sansSerifFontFamily]);
     settings->setSerifFontFamily([preferences serifFontFamily]);
     settings->setStandardFontFamily([preferences standardFontFamily]);
Index: WebKit/mac/WebView/WebPreferences.mm
===================================================================
--- WebKit/mac/WebView/WebPreferences.mm	(revision 41760)
+++ WebKit/mac/WebView/WebPreferences.mm	(working copy)
@@ -325,6 +325,7 @@
         @"1800",                        WebKitBackForwardCacheExpirationIntervalKey,
         [NSNumber numberWithBool:NO],   WebKitTabToLinksPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitPrivateBrowsingEnabledPreferenceKey,
+        [NSNumber numberWithBool:NO],   WebKitPreventUserProfilingPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitRespectStandardStyleKeyEquivalentsPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitShowsURLsInToolTipsPreferenceKey,
         @"1",                           WebKitPDFDisplayModePreferenceKey,
@@ -720,6 +721,16 @@
     return [self _boolValueForKey:WebKitPrivateBrowsingEnabledPreferenceKey];
 }
 
+- (void)setPreventUserProfiling:(BOOL)flag
+{
+    [self _setBoolValue:flag forKey:WebKitPreventUserProfilingPreferenceKey];
+}
+
+- (BOOL)preventUserProfiling
+{
+    return [self _boolValueForKey:WebKitPreventUserProfilingPreferenceKey];
+}
+
 - (void)setUsesPageCache:(BOOL)usesPageCache
 {
     [self _setBoolValue:usesPageCache forKey:WebKitUsesPageCachePreferenceKey];
Index: WebKit/qt/Api/qwebsettings.cpp
===================================================================
--- WebKit/qt/Api/qwebsettings.cpp	(revision 41760)
+++ WebKit/qt/Api/qwebsettings.cpp	(working copy)
@@ -150,6 +150,11 @@
                                       global->attributes.value(QWebSettings::PrivateBrowsingEnabled));
         settings->setPrivateBrowsingEnabled(value);
 
+
+        value = attributes.value(QWebSettings::PreventUserProfiling,
+                                 global->attributes.value(QWebSettings::PreventUserProfiling));
+        settings->setPreventUserProfiling(value);
+
         value = attributes.value(QWebSettings::JavascriptCanAccessClipboard,
                                       global->attributes.value(QWebSettings::JavascriptCanAccessClipboard));
         settings->setDOMPasteAllowed(value);
@@ -316,6 +321,9 @@
         local storage feature is enabled or not.
     \value AllowUniversalAccessFromFileUrls Specifies whether documents from file
         Urls should be granted universal access (e.g., to HTTP and HTTPS documents).
+    \value PreventUserProfiling Prevents websites from profiling and identifying
+        visitors based on their browser window properties, visited websites, date
+        and time locale information.
 */
 
 /*!
Index: WebKit/qt/Api/qwebsettings.h
===================================================================
--- WebKit/qt/Api/qwebsettings.h	(revision 41760)
+++ WebKit/qt/Api/qwebsettings.h	(working copy)
@@ -64,7 +64,8 @@
         OfflineStorageDatabaseEnabled,
         OfflineWebApplicationCacheEnabled,
         LocalStorageDatabaseEnabled,
-        AllowUniversalAccessFromFileUrls
+        AllowUniversalAccessFromFileUrls,
+        PreventUserProfiling
     };
     enum WebGraphic {
         MissingImageGraphic,
Index: WebKit/qt/Api/qwebpage.cpp
===================================================================
--- WebKit/qt/Api/qwebpage.cpp	(revision 41760)
+++ WebKit/qt/Api/qwebpage.cpp	(working copy)
@@ -2420,7 +2420,12 @@
 */
 QString QWebPage::userAgentForUrl(const QUrl& url) const
 {
+ /*torora user agent header */
     Q_UNUSED(url)
+
+    if (!QWebPage::settings() || QWebPage::settings()->PreventUserProfiling)
+      return "Mozilla/5.0 (Windows; U; Windows NT 5.1; LANG; rv:1.9.0.6) Gecko/2009011913 Firefox/3.0.6";
+    
     QString ua = QLatin1String("Mozilla/5.0 ("
 
     // Plastform
Index: WebKit/win/WebPreferences.h
===================================================================
--- WebKit/win/WebPreferences.h	(revision 41760)
+++ WebKit/win/WebPreferences.h	(working copy)
@@ -193,6 +193,12 @@
     virtual HRESULT STDMETHODCALLTYPE privateBrowsingEnabled( 
         /* [retval][out] */ BOOL* enabled);
     
+    virtual HRESULT STDMETHODCALLTYPE setPreventUserProfiling(
+        /* [in] */ BOOL enabled);
+    
+    virtual HRESULT STDMETHODCALLTYPE preventUserProfiling(
+        /* [retval][out] */ BOOL* enabled);
+    
     virtual HRESULT STDMETHODCALLTYPE setTabsToLinks( 
         /* [in] */ BOOL enabled);
     
Index: WebKit/win/WebView.cpp
===================================================================
--- WebKit/win/WebView.cpp	(revision 41760)
+++ WebKit/win/WebView.cpp	(working copy)
@@ -4055,6 +4055,11 @@
         return hr;
     settings->setPrivateBrowsingEnabled(!!enabled);
 
+    hr = preferences->preventUserProfiling(&enabled);
+    if (FAILED(hr))
+        return hr;
+    settings->setPreventUserProfiling(!!enabled);
+
     hr = preferences->sansSerifFontFamily(&str);
     if (FAILED(hr))
         return hr;
Index: WebKit/win/WebPreferenceKeysPrivate.h
===================================================================
--- WebKit/win/WebPreferenceKeysPrivate.h	(revision 41760)
+++ WebKit/win/WebPreferenceKeysPrivate.h	(working copy)
@@ -59,6 +59,7 @@
 #define WebKitBackForwardCacheExpirationIntervalKey "WebKitBackForwardCacheExpirationIntervalKey"
 #define WebKitTabToLinksPreferenceKey "WebKitTabToLinksPreferenceKey"
 #define WebKitPrivateBrowsingEnabledPreferenceKey "WebKitPrivateBrowsingEnabled"
+#define WebKitPreventUserProfilingKey "WebKitPreventUserProfiling"
 #define WebKitIconDatabaseLocationKey "WebKitIconDatabaseLocation"
 #define WebKitIconDatabaseEnabledPreferenceKey "WebKitIconDatabaseEnabled"
 #define WebKitUsesPageCachePreferenceKey "WebKitUsesPageCachePreferenceKey"
Index: WebKit/win/WebPreferences.cpp
===================================================================
--- WebKit/win/WebPreferences.cpp	(revision 41760)
+++ WebKit/win/WebPreferences.cpp	(working copy)
@@ -215,6 +215,7 @@
     CFDictionaryAddValue(defaults, CFSTR(WebKitBackForwardCacheExpirationIntervalKey), CFSTR("1800"));
     CFDictionaryAddValue(defaults, CFSTR(WebKitTabToLinksPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitPrivateBrowsingEnabledPreferenceKey), kCFBooleanFalse);
+    CFDictionaryAddValue(defaults, CFSTR(WebKitPreventUserProfilingPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitRespectStandardStyleKeyEquivalentsPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitShowsURLsInToolTipsPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitPDFDisplayModePreferenceKey), CFSTR("1"));
@@ -893,6 +894,20 @@
     return S_OK;
 }
 
+HRESULT STDMETHODCALLTYPE WebPreferences::setPreventUserProfiling(
+    /* [in] */ BOOL enabled)
+{
+    setBoolValue(CFSTR(WebKitPreventUserProfilingPreferenceKey), enabled);
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE WebPreferences::preventUserProfiling(
+    /* [retval][out] */ BOOL* enabled)
+{
+    *enabled = boolValueForKey(CFSTR(WebKitPreventUserProfilingPreferenceKey));
+    return S_OK;
+}
+
 HRESULT STDMETHODCALLTYPE WebPreferences::setTabsToLinks( 
     /* [in] */ BOOL enabled)
 {
Index: WebKit/win/Interfaces/IWebPreferences.idl
===================================================================
--- WebKit/win/Interfaces/IWebPreferences.idl	(revision 41760)
+++ WebKit/win/Interfaces/IWebPreferences.idl	(working copy)
@@ -141,6 +141,9 @@
     HRESULT setPrivateBrowsingEnabled([in] BOOL enabled);
     HRESULT privateBrowsingEnabled([out, retval] BOOL* enabled);
 
+    HRESULT setPreventUserProfiling([in] BOOL enabled);
+    HRESULT preventUserProfiling([out, retval] BOOL* enabled);
+    
     HRESULT setTabsToLinks([in] BOOL enabled);
     HRESULT tabsToLinks([out, retval] BOOL* enabled);
 
Index: WebKit/win/WebCoreSupport/WebInspectorClient.cpp
===================================================================
--- WebKit/win/WebCoreSupport/WebInspectorClient.cpp	(revision 41760)
+++ WebKit/win/WebCoreSupport/WebInspectorClient.cpp	(working copy)
@@ -145,6 +145,8 @@
         return 0;
     if (FAILED(preferences->setPrivateBrowsingEnabled(TRUE)))
         return 0;
+    if (FAILED(preferences->setPreventUserProfiling(FALSE)))
+        return 0;
     if (FAILED(preferences->setLoadsImagesAutomatically(TRUE)))
         return 0;
     if (FAILED(preferences->setAuthorAndUserStylesEnabled(TRUE)))
Index: WebCore/page/Navigator.h
===================================================================
--- WebCore/page/Navigator.h	(revision 41760)
+++ WebCore/page/Navigator.h	(working copy)
@@ -48,6 +48,8 @@
         MimeTypeArray* mimeTypes() const;
         bool cookieEnabled() const;
         bool javaEnabled() const;
+        String platform() const;
+        String vendor() const;
 
         virtual String userAgent() const;
 
Index: WebCore/page/History.cpp
===================================================================
--- WebCore/page/History.cpp	(revision 41760)
+++ WebCore/page/History.cpp	(working copy)
@@ -1,4 +1,4 @@
-/*
+  /*
  * Copyright (C) 2007 Apple Inc.  All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "History.h"
+#include "Settings.h"
 
 #include "Frame.h"
 #include "FrameLoader.h"
@@ -50,6 +51,11 @@
 {
     if (!m_frame)
         return 0;
+    Settings* settings = m_frame->settings();
+
+    if (!settings || settings->preventUserProfiling())
+      return 0;
+
     return m_frame->loader()->getHistoryLength();
 }
 
Index: WebCore/page/Settings.h
===================================================================
--- WebCore/page/Settings.h	(revision 41760)
+++ WebCore/page/Settings.h	(working copy)
@@ -112,7 +112,10 @@
 
         void setPrivateBrowsingEnabled(bool);
         bool privateBrowsingEnabled() const { return m_privateBrowsingEnabled; }
-        
+
+        void setPreventUserProfiling(bool);
+        bool preventUserProfiling() const { return m_preventUserProfiling; }
+
         void setDefaultTextEncodingName(const String&);
         const String& defaultTextEncodingName() const { return m_defaultTextEncodingName; }
 
@@ -230,6 +233,7 @@
         bool m_isJavaEnabled : 1;
         bool m_loadsImagesAutomatically : 1;
         bool m_privateBrowsingEnabled : 1;
+        bool m_preventUserProfiling : 1;
         bool m_arePluginsEnabled : 1;
         bool m_databasesEnabled : 1;
         bool m_localStorageEnabled : 1;
Index: WebCore/page/DOMWindow.cpp
===================================================================
--- WebCore/page/DOMWindow.cpp	(revision 41760)
+++ WebCore/page/DOMWindow.cpp	(working copy)
@@ -561,6 +561,11 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(page->chrome()->windowRect().height() / 50) * 50);
+
     return static_cast<int>(page->chrome()->windowRect().height());
 }
 
@@ -573,6 +578,11 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(page->chrome()->windowRect().width() / 50) * 50);
+
     return static_cast<int>(page->chrome()->windowRect().width());
 }
 
@@ -584,12 +594,18 @@
     FrameView* view = m_frame->view();
     if (!view)
         return 0;
-    
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>((static_cast<int>(view->height() / m_frame->pageZoomFactor()) / 50) * 50);
+
     return static_cast<int>(view->height() / m_frame->pageZoomFactor());
 }
 
 int DOMWindow::innerWidth() const
 {
+
     if (!m_frame)
         return 0;
 
@@ -597,11 +613,18 @@
     if (!view)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>((static_cast<int>(view->width() / 50) * 50)
+                                / m_frame->pageZoomFactor());
+
     return static_cast<int>(view->width() / m_frame->pageZoomFactor());
 }
 
 int DOMWindow::screenX() const
 {
+  /*TORORA: use settings to prevent this value being returned to javascriptcore */
     if (!m_frame)
         return 0;
 
@@ -609,6 +632,11 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(page->chrome()->windowRect().x() / 50) * 50);
+
     return static_cast<int>(page->chrome()->windowRect().x());
 }
 
@@ -621,6 +649,11 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(page->chrome()->windowRect().y() / 50) * 50);
+
     return static_cast<int>(page->chrome()->windowRect().y());
 }
 
@@ -633,6 +666,11 @@
     if (!view)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(view->scrollX() / 50) * 50);
+
     m_frame->document()->updateLayoutIgnorePendingStylesheets();
 
     return static_cast<int>(view->scrollX() / m_frame->pageZoomFactor());
@@ -640,13 +678,18 @@
 
 int DOMWindow::scrollY() const
 {
-    if (!m_frame)
+      if (!m_frame)
         return 0;
 
     FrameView* view = m_frame->view();
     if (!view)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>(static_cast<int>(view->scrollX() / 50) * 50);
+
     m_frame->document()->updateLayoutIgnorePendingStylesheets();
 
     return static_cast<int>(view->scrollY() / m_frame->pageZoomFactor());
Index: WebCore/page/NavigatorBase.cpp
===================================================================
--- WebCore/page/NavigatorBase.cpp	(revision 41760)
+++ WebCore/page/NavigatorBase.cpp	(working copy)
@@ -29,6 +29,7 @@
 
 #include "NetworkStateNotifier.h"
 #include "PlatformString.h"
+#include "Settings.h"
 
 #ifndef WEBCORE_NAVIGATOR_PLATFORM
 #if PLATFORM(MAC) && (PLATFORM(PPC) || PLATFORM(PPC64))
@@ -89,21 +90,41 @@
 
 String NavigatorBase::product() const
 {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "Gecko";
+
     return WEBCORE_NAVIGATOR_PRODUCT;
 }
 
 String NavigatorBase::productSub() const
 {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "2009011913";
+
     return WEBCORE_NAVIGATOR_PRODUCT_SUB;
 }
 
 String NavigatorBase::vendor() const
 {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "";
+
     return WEBCORE_NAVIGATOR_VENDOR;
 }
 
 String NavigatorBase::vendorSub() const
 {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "";
+
     return WEBCORE_NAVIGATOR_VENDOR_SUB;
 }
 
Index: WebCore/page/Screen.cpp
===================================================================
--- WebCore/page/Screen.cpp	(revision 41760)
+++ WebCore/page/Screen.cpp	(working copy)
@@ -34,6 +34,7 @@
 #include "Frame.h"
 #include "FrameView.h"
 #include "PlatformScreen.h"
+#include "Settings.h"
 #include "Widget.h"
 
 namespace WebCore {
@@ -52,6 +53,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<unsigned>(static_cast<int>(screenRect(m_frame->view()).height() / 50) * 50);
+
     return static_cast<unsigned>(screenRect(m_frame->view()).height());
 }
 
@@ -59,6 +66,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<unsigned>(static_cast<int>(screenRect(m_frame->view()).width() / 50) * 50);
+
     return static_cast<unsigned>(screenRect(m_frame->view()).width());
 }
 
@@ -66,6 +79,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return 24;
+
     return static_cast<unsigned>(screenDepth(m_frame->view()));
 }
 
@@ -73,6 +92,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return 24;
+
     return static_cast<unsigned>(screenDepth(m_frame->view()));
 }
 
@@ -80,6 +105,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return 0;
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).x());
 }
 
@@ -87,6 +118,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    // and https://tor-svn.freehaven.net/svn/torbutton/trunk/src/chrome/content/jshooks.js
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return 0;
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).y());
 }
 
@@ -94,6 +132,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    // and https://tor-svn.freehaven.net/svn/torbutton/trunk/src/chrome/content/jshooks.js
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>((static_cast<int>(m_frame->view()->height() / m_frame->pageZoomFactor()) / 50) * 50);
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).height());
 }
 
@@ -101,6 +146,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->preventUserProfiling())
+        return static_cast<int>((static_cast<int>(m_frame->view()->width() / m_frame->pageZoomFactor()) / 50) * 50);
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).width());
 }
 
Index: WebCore/page/Navigator.cpp
===================================================================
--- WebCore/page/Navigator.cpp	(revision 41760)
+++ WebCore/page/Navigator.cpp	(working copy)
@@ -87,6 +87,11 @@
 {
     if (!m_frame)
         return String();
+    Settings* settings = m_frame->settings();
+
+    if (!settings || settings->preventUserProfiling())
+      return "5.0 (Windows; LANG)";
+
     String appVersion = NavigatorBase::appVersion();
     if (shouldHideFourDot(m_frame))
         appVersion.replace("4.", "4_");
@@ -95,6 +100,11 @@
 
 String Navigator::language() const
 {
+    Settings* settings = m_frame->settings();
+
+    if (!settings || settings->preventUserProfiling())
+      return "en-us, en";
+
     return defaultLanguage();
 }
 
@@ -102,7 +112,12 @@
 {
     if (!m_frame)
         return String();
-        
+
+    Settings* settings = m_frame->settings();
+
+    if (!settings || settings->preventUserProfiling())
+      return "Mozilla/5.0 (Windows; U; Windows NT 5.1; LANG; rv:1.9.0.6) Gecko/2009011913 Firefox/3.0.6";
+
     // If the frame is already detached, FrameLoader::userAgent may malfunction, because it calls a client method
     // that uses frame's WebView (at least, in Mac WebKit).
     if (!m_frame->page())
@@ -144,6 +159,62 @@
     return m_frame->settings()->isJavaEnabled();
 }
 
+String Navigator::platform() const
+{
+//     if (!m_frame || !m_frame->settings())
+//         return String();
+
+    if (m_frame->settings()->preventUserProfiling())
+      return "Win32";
+
+    return NavigatorBase::platform();
+}
+
+// String Navigator::appCodeName() const
+// {
+//     return "Mozilla";
+// }
+// 
+// String Navigator::product() const
+// {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "Gecko";
+// 
+//     return WEBCORE_NAVIGATOR_PRODUCT;
+// }
+// 
+// String Navigator::productSub() const
+// {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "2009011913";
+// 
+//     return WEBCORE_NAVIGATOR_PRODUCT_SUB;
+// }
+// 
+String Navigator::vendor() const
+{
+    Settings* settings = m_frame->settings();
+
+    if (!settings || settings->preventUserProfiling())
+      return "";
+
+    return NavigatorBase::vendor();
+}
+
+// String Navigator::vendorSub() const
+// {
+//     Settings* settings = m_frame->settings();
+// 
+//     if (!settings || settings->preventUserProfiling())
+//       return "";
+// 
+//     return WEBCORE_NAVIGATOR_VENDOR_SUB;
+// }
+
 Geolocation* Navigator::geolocation() const
 {
     if (!m_geolocation)
Index: WebCore/page/Settings.cpp
===================================================================
--- WebCore/page/Settings.cpp	(revision 41760)
+++ WebCore/page/Settings.cpp	(working copy)
@@ -56,6 +56,7 @@
     , m_isJavaEnabled(false)
     , m_loadsImagesAutomatically(false)
     , m_privateBrowsingEnabled(false)
+    , m_preventUserProfiling(false)
     , m_arePluginsEnabled(false)
     , m_databasesEnabled(false)
     , m_localStorageEnabled(false)
@@ -229,6 +230,11 @@
     m_privateBrowsingEnabled = privateBrowsingEnabled;
 }
 
+void Settings::setPreventUserProfiling(bool preventUserProfiling)
+{
+    m_preventUserProfiling = preventUserProfiling;
+}
+
 void Settings::setJavaScriptCanOpenWindowsAutomatically(bool javaScriptCanOpenWindowsAutomatically)
 {
     m_javaScriptCanOpenWindowsAutomatically = javaScriptCanOpenWindowsAutomatically;
Index: WebCore/platform/graphics/qt/GraphicsContextQt.cpp
===================================================================
--- WebCore/platform/graphics/qt/GraphicsContextQt.cpp	(revision 41760)
+++ WebCore/platform/graphics/qt/GraphicsContextQt.cpp	(working copy)
@@ -473,6 +473,7 @@
         p->restore();
     }
 
+#if QT_VERSION >= 0x040500
     int patWidth = 0;
     switch (style) {
         case NoStroke:
@@ -488,7 +489,6 @@
 
     if (patWidth) {
         p->save();
-
         // Do a rect fill of our endpoints.  This ensures we always have the
         // appearance of being a border.  We then draw the actual dotted/dashed line.
         if (isVerticalLine) {
@@ -498,7 +498,6 @@
             p->fillRect(FloatRect(p1.x() - width, p1.y() - width / 2, width, width), color);
             p->fillRect(FloatRect(p2.x(), p2.y() - width / 2, width, width), color);
         }
-
         // Example: 80 pixels with a width of 30 pixels.
         // Remainder is 20.  The maximum pixels of line we could paint
         // will be 50 pixels.
@@ -537,12 +536,14 @@
         pen.setDashOffset(patternOffset / width);
         p->setPen(pen);
     }
+#endif
 
     p->drawLine(p1, p2);
 
+#if QT_VERSION >= 0x040500
     if (patWidth)
         p->restore();
-
+#endif
     p->setRenderHint(QPainter::Antialiasing, antiAlias);
 }
 
Index: WebCore/bindings/js/JSDOMWindowBase.h
===================================================================
--- WebCore/bindings/js/JSDOMWindowBase.h	(revision 41760)
+++ WebCore/bindings/js/JSDOMWindowBase.h	(working copy)
@@ -81,6 +81,8 @@
 
         virtual bool shouldInterruptScript() const;
 
+        virtual bool shouldPreventUserProfiling() const;
+
         bool allowsAccessFrom(JSC::ExecState*) const;
         bool allowsAccessFromNoErrorMessage(JSC::ExecState*) const;
         bool allowsAccessFrom(JSC::ExecState*, String& message) const;
Index: WebCore/bindings/js/JSDOMWindowBase.cpp
===================================================================
--- WebCore/bindings/js/JSDOMWindowBase.cpp	(revision 41760)
+++ WebCore/bindings/js/JSDOMWindowBase.cpp	(working copy)
@@ -683,7 +683,7 @@
     Settings* settings = frame->settings();
     if (!settings)
         return;
-    
+
     if (settings->privateBrowsingEnabled())
         return;
 
@@ -730,6 +730,15 @@
     return page->chrome()->shouldInterruptJavaScript();
 }
 
+bool JSDOMWindowBase::shouldPreventUserProfiling() const
+{
+    ASSERT(impl()->frame());
+    Settings* settings = impl()->frame()->settings();
+    if (!settings || settings->preventUserProfiling())
+      return true;
+    return false;
+}
+
 void JSDOMWindowBase::clearHelperObjectProperties()
 {
     setCurrentEvent(0);
Index: WebCore/css/CSSStyleSelector.cpp
===================================================================
--- WebCore/css/CSSStyleSelector.cpp	(revision 41760)
+++ WebCore/css/CSSStyleSelector.cpp	(working copy)
@@ -875,6 +875,10 @@
     if (!page)
         return PseudoLink;
 
+    Settings* settings = m_document->settings();
+    if (!settings || settings->preventUserProfiling())
+        return PseudoLink;
+
     m_linksCheckedForVisitedState.add(hash);
     return page->group().isLinkVisited(hash) ? PseudoVisited : PseudoLink;
 }
@@ -5840,6 +5844,11 @@
 
 void CSSStyleSelector::SelectorChecker::allVisitedStateChanged()
 {
+
+    Settings* settings = m_document->settings();
+    if (!settings || settings->preventUserProfiling())
+        return;
+
     if (m_linksCheckedForVisitedState.isEmpty())
         return;
     for (Node* node = m_document; node; node = node->traverseNextNode()) {
@@ -5850,6 +5859,11 @@
 
 void CSSStyleSelector::SelectorChecker::visitedStateChanged(LinkHash visitedHash)
 {
+
+    Settings* settings = m_document->settings();
+    if (!settings || settings->preventUserProfiling())
+        return;
+
     if (!m_linksCheckedForVisitedState.contains(visitedHash))
         return;
     for (Node* node = m_document; node; node = node->traverseNextNode()) {
Index: WebKitTools/Scripts/webkitdirs.pm
===================================================================
--- WebKitTools/Scripts/webkitdirs.pm	(revision 41760)
+++ WebKitTools/Scripts/webkitdirs.pm	(working copy)
@@ -1026,7 +1026,7 @@
         return "nmake";
     }
 
-    return "make";
+    return "make -j4";
 }
 
 sub autotoolsFlag($$)
