Index: JavaScriptCore/runtime/DatePrototype.cpp
===================================================================
--- JavaScriptCore/runtime/DatePrototype.cpp	(revision 41537)
+++ JavaScriptCore/runtime/DatePrototype.cpp	(working copy)
@@ -776,6 +776,7 @@
 
 JSValuePtr dateProtoFuncGetTimezoneOffset(ExecState* exec, JSObject*, JSValuePtr thisValue, const ArgList&)
 {
+  /*TORORA: need to intercept gettimezoneoffset here */
     if (!thisValue.isObject(&DateInstance::info))
         return throwError(exec, TypeError);
 
Index: WebKit/mac/WebCoreSupport/WebInspectorClient.mm
===================================================================
--- WebKit/mac/WebCoreSupport/WebInspectorClient.mm	(revision 41537)
+++ WebKit/mac/WebCoreSupport/WebInspectorClient.mm	(working copy)
@@ -167,6 +167,7 @@
     WebPreferences *preferences = [[WebPreferences alloc] init];
     [preferences setAutosaves:NO];
     [preferences setPrivateBrowsingEnabled:YES];
+    [preferences setConcealVisitedLinks:NO];
     [preferences setLoadsImagesAutomatically:YES];
     [preferences setAuthorAndUserStylesEnabled:YES];
     [preferences setJavaScriptEnabled:YES];
Index: WebKit/mac/WebView/WebPreferences.h
===================================================================
--- WebKit/mac/WebView/WebPreferences.h	(revision 41537)
+++ WebKit/mac/WebView/WebPreferences.h	(working copy)
@@ -366,6 +366,19 @@
 - (BOOL)privateBrowsingEnabled;
 
 /*!
+    @method setConcealVisitedLinks:
+    @param flag 
+    @abstract Prevents websites from using the 'visited' property
+        in CSS stylesheets to gather information on your browsing history..
+ */
+- (void)setConcealVisitedLinks:(BOOL)flag;
+
+/*!
+    @method privateBrowsingEnabled
+ */
+- (BOOL)concealVisitedLinks;
+
+/*!
     @method setTabsToLinks:
     @param flag 
     @abstract If tabsToLinks is YES, the tab key will focus links and form controls. 
Index: WebKit/mac/WebView/WebPreferenceKeysPrivate.h
===================================================================
--- WebKit/mac/WebView/WebPreferenceKeysPrivate.h	(revision 41537)
+++ WebKit/mac/WebView/WebPreferenceKeysPrivate.h	(working copy)
@@ -60,6 +60,7 @@
 #define WebKitBackForwardCacheExpirationIntervalKey @"WebKitBackForwardCacheExpirationIntervalKey"
 #define WebKitTabToLinksPreferenceKey @"WebKitTabToLinksPreferenceKey"
 #define WebKitPrivateBrowsingEnabledPreferenceKey @"WebKitPrivateBrowsingEnabled"
+#define WebKitConcealVisitedLinksPreferenceKey @"WebKitConcealVisitedLinks"
 #define WebContinuousSpellCheckingEnabled @"WebContinuousSpellCheckingEnabled"
 #define WebGrammarCheckingEnabled @"WebGrammarCheckingEnabled"
 #define WebKitDOMPasteAllowedPreferenceKey @"WebKitDOMPasteAllowedPreferenceKey"
Index: WebKit/mac/WebView/WebView.mm
===================================================================
--- WebKit/mac/WebView/WebView.mm	(revision 41537)
+++ WebKit/mac/WebView/WebView.mm	(working copy)
@@ -1325,6 +1325,7 @@
     settings->setDatabasesEnabled([preferences databasesEnabled]);
     settings->setLocalStorageEnabled([preferences localStorageEnabled]);
     settings->setPrivateBrowsingEnabled([preferences privateBrowsingEnabled]);
+    settings->setConcealVisitedLinks([preferences concealVisitedLinks]);
     settings->setSansSerifFontFamily([preferences sansSerifFontFamily]);
     settings->setSerifFontFamily([preferences serifFontFamily]);
     settings->setStandardFontFamily([preferences standardFontFamily]);
Index: WebKit/mac/WebView/WebPreferences.mm
===================================================================
--- WebKit/mac/WebView/WebPreferences.mm	(revision 41537)
+++ WebKit/mac/WebView/WebPreferences.mm	(working copy)
@@ -325,6 +325,7 @@
         @"1800",                        WebKitBackForwardCacheExpirationIntervalKey,
         [NSNumber numberWithBool:NO],   WebKitTabToLinksPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitPrivateBrowsingEnabledPreferenceKey,
+        [NSNumber numberWithBool:NO],   WebKitConcealVisitedLinksPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitRespectStandardStyleKeyEquivalentsPreferenceKey,
         [NSNumber numberWithBool:NO],   WebKitShowsURLsInToolTipsPreferenceKey,
         @"1",                           WebKitPDFDisplayModePreferenceKey,
@@ -720,6 +721,16 @@
     return [self _boolValueForKey:WebKitPrivateBrowsingEnabledPreferenceKey];
 }
 
+- (void)setConcealVisitedLinks:(BOOL)flag
+{
+    [self _setBoolValue:flag forKey:WebKitConcealVisitedLinksPreferenceKey];
+}
+
+- (BOOL)concealVisitedLinks
+{
+    return [self _boolValueForKey:WebKitConcealVisitedLinksPreferenceKey];
+}
+
 - (void)setUsesPageCache:(BOOL)usesPageCache
 {
     [self _setBoolValue:usesPageCache forKey:WebKitUsesPageCachePreferenceKey];
Index: WebKit/qt/Api/qwebsettings.cpp
===================================================================
--- WebKit/qt/Api/qwebsettings.cpp	(revision 41537)
+++ WebKit/qt/Api/qwebsettings.cpp	(working copy)
@@ -150,6 +150,14 @@
                                       global->attributes.value(QWebSettings::PrivateBrowsingEnabled));
         settings->setPrivateBrowsingEnabled(value);
 
+        value = attributes.value(QWebSettings::ConcealVisitedLinks,
+                                      global->attributes.value(QWebSettings::ConcealVisitedLinks));
+        settings->setConcealVisitedLinks(value);
+
+        value = attributes.value(QWebSettings::JSPrivateWindowProperties,
+                                 global->attributes.value(QWebSettings::JSPrivateWindowProperties));
+        settings->setJSPrivateWindowProperties(value);
+
         value = attributes.value(QWebSettings::JavascriptCanAccessClipboard,
                                       global->attributes.value(QWebSettings::JavascriptCanAccessClipboard));
         settings->setDOMPasteAllowed(value);
@@ -314,8 +322,12 @@
         web application cache feature is enabled or not.
     \value LocalStorageDatabaseEnabled Specifies whether support for the HTML 5
         local storage feature is enabled or not.
+    \value ConcealVisitedLinks Prevents websites from using the 'visited' property
+        in CSS stylesheets to gather information on your browsing history.
     \value AllowUniversalAccessFromFileUrls Specifies whether documents from file
         Urls should be granted universal access (e.g., to HTTP and HTTPS documents).
+    \value JSPrivateWindowProperties Prevents websites from identifying anonymous
+        visitors based on their browser window properties.
 */
 
 /*!
Index: WebKit/qt/Api/qwebsettings.h
===================================================================
--- WebKit/qt/Api/qwebsettings.h	(revision 41537)
+++ WebKit/qt/Api/qwebsettings.h	(working copy)
@@ -64,7 +64,9 @@
         OfflineStorageDatabaseEnabled,
         OfflineWebApplicationCacheEnabled,
         LocalStorageDatabaseEnabled,
-        AllowUniversalAccessFromFileUrls
+        AllowUniversalAccessFromFileUrls,
+        ConcealVisitedLinks,
+        JSPrivateWindowProperties
     };
     enum WebGraphic {
         MissingImageGraphic,
Index: WebKit/win/WebPreferences.h
===================================================================
--- WebKit/win/WebPreferences.h	(revision 41537)
+++ WebKit/win/WebPreferences.h	(working copy)
@@ -193,6 +193,12 @@
     virtual HRESULT STDMETHODCALLTYPE privateBrowsingEnabled( 
         /* [retval][out] */ BOOL* enabled);
     
+    virtual HRESULT STDMETHODCALLTYPE setConcealVisitedLinks( 
+        /* [in] */ BOOL enabled);
+    
+    virtual HRESULT STDMETHODCALLTYPE concealVisitedLinks( 
+        /* [retval][out] */ BOOL* enabled);
+    
     virtual HRESULT STDMETHODCALLTYPE setTabsToLinks( 
         /* [in] */ BOOL enabled);
     
Index: WebKit/win/WebPreferenceKeysPrivate.h
===================================================================
--- WebKit/win/WebPreferenceKeysPrivate.h	(revision 41537)
+++ WebKit/win/WebPreferenceKeysPrivate.h	(working copy)
@@ -59,6 +59,7 @@
 #define WebKitBackForwardCacheExpirationIntervalKey "WebKitBackForwardCacheExpirationIntervalKey"
 #define WebKitTabToLinksPreferenceKey "WebKitTabToLinksPreferenceKey"
 #define WebKitPrivateBrowsingEnabledPreferenceKey "WebKitPrivateBrowsingEnabled"
+#define WebKitConcealVisitedLinksKey "WebKitConcealVisitedLinks"
 #define WebKitIconDatabaseLocationKey "WebKitIconDatabaseLocation"
 #define WebKitIconDatabaseEnabledPreferenceKey "WebKitIconDatabaseEnabled"
 #define WebKitUsesPageCachePreferenceKey "WebKitUsesPageCachePreferenceKey"
Index: WebKit/win/WebView.cpp
===================================================================
--- WebKit/win/WebView.cpp	(revision 41537)
+++ WebKit/win/WebView.cpp	(working copy)
@@ -4055,6 +4055,11 @@
         return hr;
     settings->setPrivateBrowsingEnabled(!!enabled);
 
+    hr = preferences->concealVisitedLinks(&enabled);
+    if (FAILED(hr))
+        return hr;
+    settings->setConcealVisitedLinks(!!enabled);
+
     hr = preferences->sansSerifFontFamily(&str);
     if (FAILED(hr))
         return hr;
Index: WebKit/win/WebPreferences.cpp
===================================================================
--- WebKit/win/WebPreferences.cpp	(revision 41537)
+++ WebKit/win/WebPreferences.cpp	(working copy)
@@ -215,6 +215,7 @@
     CFDictionaryAddValue(defaults, CFSTR(WebKitBackForwardCacheExpirationIntervalKey), CFSTR("1800"));
     CFDictionaryAddValue(defaults, CFSTR(WebKitTabToLinksPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitPrivateBrowsingEnabledPreferenceKey), kCFBooleanFalse);
+    CFDictionaryAddValue(defaults, CFSTR(WebKitConcealVisitedLinksPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitRespectStandardStyleKeyEquivalentsPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitShowsURLsInToolTipsPreferenceKey), kCFBooleanFalse);
     CFDictionaryAddValue(defaults, CFSTR(WebKitPDFDisplayModePreferenceKey), CFSTR("1"));
@@ -893,6 +894,20 @@
     return S_OK;
 }
 
+HRESULT STDMETHODCALLTYPE WebPreferences::setConcealVisitedLinks( 
+    /* [in] */ BOOL enabled)
+{
+    setBoolValue(CFSTR(WebKitConcealVisitedLinksPreferenceKey), enabled);
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE WebPreferences::concealVisitedLinks( 
+    /* [retval][out] */ BOOL* enabled)
+{
+    *enabled = boolValueForKey(CFSTR(WebKitConcealVisitedLinksPreferenceKey));
+    return S_OK;
+}
+
 HRESULT STDMETHODCALLTYPE WebPreferences::setTabsToLinks( 
     /* [in] */ BOOL enabled)
 {
Index: WebKit/win/Interfaces/IWebPreferences.idl
===================================================================
--- WebKit/win/Interfaces/IWebPreferences.idl	(revision 41537)
+++ WebKit/win/Interfaces/IWebPreferences.idl	(working copy)
@@ -141,6 +141,9 @@
     HRESULT setPrivateBrowsingEnabled([in] BOOL enabled);
     HRESULT privateBrowsingEnabled([out, retval] BOOL* enabled);
 
+    HRESULT setConcealVisitedLinks([in] BOOL enabled);
+    HRESULT concealVisitedLinks([out, retval] BOOL* enabled);
+    
     HRESULT setTabsToLinks([in] BOOL enabled);
     HRESULT tabsToLinks([out, retval] BOOL* enabled);
 
Index: WebKit/win/WebCoreSupport/WebInspectorClient.cpp
===================================================================
--- WebKit/win/WebCoreSupport/WebInspectorClient.cpp	(revision 41537)
+++ WebKit/win/WebCoreSupport/WebInspectorClient.cpp	(working copy)
@@ -145,6 +145,8 @@
         return 0;
     if (FAILED(preferences->setPrivateBrowsingEnabled(TRUE)))
         return 0;
+    if (FAILED(preferences->setConcealVisitedLinks(FALSE)))
+        return 0;
     if (FAILED(preferences->setLoadsImagesAutomatically(TRUE)))
         return 0;
     if (FAILED(preferences->setAuthorAndUserStylesEnabled(TRUE)))
Index: WebCore/page/Settings.h
===================================================================
--- WebCore/page/Settings.h	(revision 41537)
+++ WebCore/page/Settings.h	(working copy)
@@ -112,7 +112,13 @@
 
         void setPrivateBrowsingEnabled(bool);
         bool privateBrowsingEnabled() const { return m_privateBrowsingEnabled; }
-        
+
+        void setConcealVisitedLinks(bool);
+        bool concealVisitedLinks() const { return m_concealVisitedLinks; }
+
+        void setJSPrivateWindowProperties(bool);
+        bool jSPrivateWindowProperties() const { return m_jSPrivateWindowProperties; }
+
         void setDefaultTextEncodingName(const String&);
         const String& defaultTextEncodingName() const { return m_defaultTextEncodingName; }
 
@@ -230,6 +236,8 @@
         bool m_isJavaEnabled : 1;
         bool m_loadsImagesAutomatically : 1;
         bool m_privateBrowsingEnabled : 1;
+        bool m_concealVisitedLinks : 1;
+        bool m_jSPrivateWindowProperties : 1;
         bool m_arePluginsEnabled : 1;
         bool m_databasesEnabled : 1;
         bool m_localStorageEnabled : 1;
Index: WebCore/page/DOMWindow.cpp
===================================================================
--- WebCore/page/DOMWindow.cpp	(revision 41537)
+++ WebCore/page/DOMWindow.cpp	(working copy)
@@ -561,6 +561,12 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>(page->chrome()->windowRect().height() -
+                                (static_cast<int>(page->chrome()->windowRect().height()) % 50));
+
     return static_cast<int>(page->chrome()->windowRect().height());
 }
 
@@ -573,6 +579,12 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>(page->chrome()->windowRect().width() -
+                                (static_cast<int>(page->chrome()->windowRect().width()) % 50));
+
     return static_cast<int>(page->chrome()->windowRect().width());
 }
 
@@ -584,12 +596,19 @@
     FrameView* view = m_frame->view();
     if (!view)
         return 0;
-    
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>((view->height() - (view->height() % 50))
+                                / m_frame->pageZoomFactor());
+
     return static_cast<int>(view->height() / m_frame->pageZoomFactor());
 }
 
 int DOMWindow::innerWidth() const
 {
+
     if (!m_frame)
         return 0;
 
@@ -597,11 +616,18 @@
     if (!view)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>((view->width() - (view->width() % 50))
+                                / m_frame->pageZoomFactor());
+
     return static_cast<int>(view->width() / m_frame->pageZoomFactor());
 }
 
 int DOMWindow::screenX() const
 {
+  /*TORORA: use settings to prevent this value being returned to javascriptcore */
     if (!m_frame)
         return 0;
 
@@ -609,6 +635,12 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>(page->chrome()->windowRect().x() -
+                                (static_cast<int>(page->chrome()->windowRect().x()) % 50));
+
     return static_cast<int>(page->chrome()->windowRect().x());
 }
 
@@ -621,6 +653,12 @@
     if (!page)
         return 0;
 
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<int>(page->chrome()->windowRect().y() -
+                                (static_cast<int>(page->chrome()->windowRect().y()) % 50));
+
     return static_cast<int>(page->chrome()->windowRect().y());
 }
 
Index: WebCore/page/Screen.cpp
===================================================================
--- WebCore/page/Screen.cpp	(revision 41537)
+++ WebCore/page/Screen.cpp	(working copy)
@@ -34,6 +34,7 @@
 #include "Frame.h"
 #include "FrameView.h"
 #include "PlatformScreen.h"
+#include "Settings.h"
 #include "Widget.h"
 
 namespace WebCore {
@@ -52,6 +53,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenRect(m_frame->view()).height() -
+                                    (static_cast<unsigned>(screenRect(m_frame->view()).height()) % 50));
+
     return static_cast<unsigned>(screenRect(m_frame->view()).height());
 }
 
@@ -59,6 +67,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenRect(m_frame->view()).width() -
+                                    (static_cast<unsigned>(screenRect(m_frame->view()).width()) % 50));
+
     return static_cast<unsigned>(screenRect(m_frame->view()).width());
 }
 
@@ -66,6 +81,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return 24;
+
     return static_cast<unsigned>(screenDepth(m_frame->view()));
 }
 
@@ -73,6 +94,12 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return 24;
+
     return static_cast<unsigned>(screenDepth(m_frame->view()));
 }
 
@@ -80,6 +107,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenAvailableRect(m_frame->view()).x() -
+                                    (static_cast<unsigned>(screenAvailableRect(m_frame->view()).x()) % 50));
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).x());
 }
 
@@ -87,6 +121,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenAvailableRect(m_frame->view()).y() -
+                                    (static_cast<unsigned>(screenAvailableRect(m_frame->view()).y()) % 50));
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).y());
 }
 
@@ -94,6 +135,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenAvailableRect(m_frame->view()).height() - 
+                                    (static_cast<unsigned>(screenAvailableRect(m_frame->view()).height()) % 50));
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).height());
 }
 
@@ -101,6 +149,13 @@
 {
     if (!m_frame)
         return 0;
+
+    // See https://www.torproject.org/torbutton/design/#id3016866
+    Settings* settings = m_frame->settings();
+    if (!settings || settings->jSPrivateWindowProperties())
+        return static_cast<unsigned>(screenAvailableRect(m_frame->view()).width() - 
+                                    (static_cast<unsigned>(screenAvailableRect(m_frame->view()).width()) % 50));
+
     return static_cast<unsigned>(screenAvailableRect(m_frame->view()).width());
 }
 
Index: WebCore/page/Settings.cpp
===================================================================
--- WebCore/page/Settings.cpp	(revision 41537)
+++ WebCore/page/Settings.cpp	(working copy)
@@ -56,6 +56,8 @@
     , m_isJavaEnabled(false)
     , m_loadsImagesAutomatically(false)
     , m_privateBrowsingEnabled(false)
+    , m_concealVisitedLinks(false)
+    , m_jSPrivateWindowProperties(false)
     , m_arePluginsEnabled(false)
     , m_databasesEnabled(false)
     , m_localStorageEnabled(false)
@@ -229,6 +231,16 @@
     m_privateBrowsingEnabled = privateBrowsingEnabled;
 }
 
+void Settings::setConcealVisitedLinks(bool concealVisitedLinks)
+{
+    m_concealVisitedLinks = concealVisitedLinks;
+}
+
+void Settings::setJSPrivateWindowProperties(bool jSPrivateWindowProperties)
+{
+    m_jSPrivateWindowProperties = jSPrivateWindowProperties;
+}
+
 void Settings::setJavaScriptCanOpenWindowsAutomatically(bool javaScriptCanOpenWindowsAutomatically)
 {
     m_javaScriptCanOpenWindowsAutomatically = javaScriptCanOpenWindowsAutomatically;
Index: WebCore/bindings/js/JSDOMWindowBase.cpp
===================================================================
--- WebCore/bindings/js/JSDOMWindowBase.cpp	(revision 41537)
+++ WebCore/bindings/js/JSDOMWindowBase.cpp	(working copy)
@@ -683,7 +683,7 @@
     Settings* settings = frame->settings();
     if (!settings)
         return;
-    
+
     if (settings->privateBrowsingEnabled())
         return;
 
Index: WebCore/css/CSSStyleSelector.cpp
===================================================================
--- WebCore/css/CSSStyleSelector.cpp	(revision 41537)
+++ WebCore/css/CSSStyleSelector.cpp	(working copy)
@@ -875,6 +875,10 @@
     if (!page)
         return PseudoLink;
 
+    Settings* settings = m_document->settings();
+    if (!settings || settings->concealVisitedLinks())
+        return PseudoLink;
+
     m_linksCheckedForVisitedState.add(hash);
     return page->group().isLinkVisited(hash) ? PseudoVisited : PseudoLink;
 }
@@ -5837,6 +5841,11 @@
 
 void CSSStyleSelector::SelectorChecker::allVisitedStateChanged()
 {
+
+    Settings* settings = m_document->settings();
+    if (!settings || settings->concealVisitedLinks())
+        return;
+
     if (m_linksCheckedForVisitedState.isEmpty())
         return;
     for (Node* node = m_document; node; node = node->traverseNextNode()) {
@@ -5847,6 +5856,11 @@
 
 void CSSStyleSelector::SelectorChecker::visitedStateChanged(LinkHash visitedHash)
 {
+
+    Settings* settings = m_document->settings();
+    if (!settings || settings->concealVisitedLinks())
+        return;
+
     if (!m_linksCheckedForVisitedState.contains(visitedHash))
         return;
     for (Node* node = m_document; node; node = node->traverseNextNode()) {
Index: WebKitTools/Scripts/webkitdirs.pm
===================================================================
--- WebKitTools/Scripts/webkitdirs.pm	(revision 41537)
+++ WebKitTools/Scripts/webkitdirs.pm	(working copy)
@@ -951,7 +951,7 @@
         return "nmake";
     }
 
-    return "make";
+    return "make -j4";
 }
 
 sub autotoolsFlag($$)
